AutoDefineAddon/libs/cardbuilder.py

```diff
@@ -0,0 +1,367 @@
+                to_print = to_print.replace(" ; ", "; ").strip()
+                to_print += "\n<br>"
+
+                # add verb/noun/adjective
+                if last_functional_label != definition.tail:
+                    to_print = definition.tail + " " + to_print
+                last_functional_label = definition.tail
+                to_return += to_print
+
+            # final cleanup of <sx> tag bs
+            to_return = to_return.replace(".</b> ; ", ".</b> ")  # <sx> as first definition after "n. " or "v. "
+            to_return = to_return.replace("\n; ", "\n")  # <sx> as first definition after newline
+            self._card.fields[settings.DEFINITION_FIELD] = to_return
+
+    def addTranscription(self):
+        valid_entries = self._card.entries
+        # extract phonetic transcriptions for each entry and label them by part of speech
+        all_transcriptions = []
+        for entry in valid_entries:
+            if entry.find("pr") is not None:
+                phonetic_transcription = entry.find("pr").text
+
+                part_of_speech = entry.find("fl").text
+                part_of_speech = _abbreviate_part_of_speech(part_of_speech)
+
+                row = f'<b>{part_of_speech}</b> \\{phonetic_transcription}\\'
+                all_transcriptions.append(row)
+
+        to_print = "<br>".join(all_transcriptions)
+
+        self._card.fields[settings.PHONETIC_TRANSCRIPTION_FIELD] = to_print
+
+    def addPronunciation(self):
+        valid_entries = self._card.entries
+        # Parse all unique pronunciations, and convert them to URLs as per http://goo.gl/nL0vte
+        all_sounds = []
+        for entry in valid_entries:
+            for wav in entry.findall("sound/wav"):
+                raw_wav = wav.text
+                # API-specific URL conversions
+                if raw_wav[:3] == "bix":
+                    mid_url = "bix"
+                elif raw_wav[:2] == "gg":
+                    mid_url = "gg"
+                elif raw_wav[:1].isdigit():
+                    mid_url = "number"
+                else:
+                    mid_url = raw_wav[:1]
+                wav_url = "http://media.merriam-webster.com/soundc11/" + mid_url + "/" + raw_wav
+                all_sounds.append(wav_url.strip())
+
+        # We want to make this a non-duplicate list, so that we only get unique sound files.
+        all_sounds = list(dict.fromkeys(all_sounds))
+        final_pronounce_index = settings.PRONUNCIATION_FIELD
+        if mw and False:
+            fields = mw.col.models.fieldNames(editor.note.model())
+            for field in fields:
+                if '🔊' in field:
+                    final_pronounce_index = fields.index(field)
+                    break
+
+        to_print = ''.join(all_sounds)
+
+        self._card.fields[settings.PRONUNCIATION_FIELD] = to_print
+
+class SpanishCardBuilder(CardBuilder):
```

Comment:
It looks like we're using the JSON api here instead of the xml one -- I am assuming that's why there is much less code re-use between the two. Can you describe why to make this choice? I'm concerned this is duplicating a significant amount of parsing code, which can really easily start to drift. Did you try to simply re-use the existing code to parse the xml-based responses?

Reply:
Thanks for the reply!

The thight is that XML API is deprecated and there is no way to get a key for it. At least, I didn't find one. If you know how to get a key for XML version of Spanish dictionary, please share, I will definitely give it a try.

I want to extend AutoDefine to work with Spanish language and the fastest way I came up with is to implement a JSON parser for merriam webster dictionary API. I guess that the parser I implemented in SpanishCardBuilder should work with other dictionaries but I need to do more testing to confirm it.

Reply:
Yeah I don't mean to block you too much but with this large of a refactoring I would prefer if we switch all the code over to the same base parsing code, ideally JSON. Happy to review that if you add commits to this PR!

Reply:
It is not a problem. I am glad to help if it makes the addon better.
Does it make sense to keep XML as an option? We can say that XML is deprecated and will be removed in future.
It might be useful if JSON parser introduces some bug. 
In that case, a user can switch back to XML instead of re-installing an older version of the extention.

Reply:
My bias would generally be to remove that support altogether as I think the option will almost never be used by users & will mean a higher volume of code to maintain. If M-W decides to actually break the JSON API for a long period of time, we should just push out a full add-on update to go back to XML. We can always find that code in the git history. But if you find a simple way to keep it we can do so -- I'll let you make the call.